pallets/amm/src/lib.rs

```diff
@@ -103,21 +119,149 @@ mod pallet {
 
     /// Accounts that deposits and withdraw assets in one or more pools
     #[pallet::storage]
-    pub type LiquidityProviders<T: Config<I>, I: 'static = ()> = StorageDoubleMap<
+    #[pallet::getter(fn liquidity_providers)]
+    pub type LiquidityProviders<T: Config<I>, I: 'static = ()> = StorageNMap<
         _,
-        Blake2_128Concat,
-        T::PoolId,
-        Blake2_128Concat,
-        T::AccountId,
-        LiquidityProviderAmounts,
+        (
+            NMapKey<Blake2_128Concat, T::AccountId>,
+            NMapKey<Blake2_128Concat, CurrencyId>,
+            NMapKey<Blake2_128Concat, CurrencyId>,
+        ),
+        PoolLiquidityAmount,
+        ValueQuery,
+        GetDefault,
     >;
 
     /// A bag of liquidity composed by two different assets
     #[pallet::storage]
-    pub type Pools<T: Config<I>, I: 'static = ()> =
-        StorageMap<_, Blake2_128Concat, T::PoolId, Pool>;
+    #[pallet::getter(fn pools)]
+    pub type Pools<T: Config<I>, I: 'static = ()> = StorageDoubleMap<
+        _,
+        Twox64Concat,
+        CurrencyId,
+        Twox64Concat,
+        CurrencyId,
+        PoolLiquidityAmount,
+        ValueQuery,
+    >;
 
-    /// Auxiliary storage used to track pool ids
-    #[pallet::storage]
-    pub type PoolsCounter<T: Config<I>, I: 'static = ()> = StorageValue<_, T::PoolId, ValueQuery>;
+    #[pallet::call]
+    impl<T: Config<I>, I: 'static> Pallet<T, I> {
+        /// Allow users to add liquidity to a given pool
+        ///
+        /// - `pool`: Currency pool, in which liquidity will be added
+        /// - `liquidity_amounts`: Liquidity amounts to be added in pool
+        #[pallet::weight(10_000)]
+        pub fn add_liquidity(
+            origin: OriginFor<T>,
+            pool: (CurrencyId, CurrencyId),
+            liquidity_amounts: (Balance, Balance),
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            let sorted_assets = &Self::get_upper_currency(pool.0, pool.1);
+
+            let base_asset = sorted_assets.0;
+            let quote_asset = sorted_assets.1;
+            let base_amount = liquidity_amounts.0;
+            let quote_amount = liquidity_amounts.1;
+
+            ensure!(
+                !Pools::<T, I>::contains_key(base_asset, &quote_asset),
+                Error::<T, I>::PoolAlreadyExists
+            );
+
+            ensure!(
+                !LiquidityProviders::<T, I>::contains_key((who.clone(), base_asset, &quote_asset)),
+                Error::<T, I>::LiquidityProviderAlreadyExists
+            );
```

Comment:
If the liquidity provider already exists we should be able to increase their liquidity

Reply:
A liquidity provider should be able to deposit tokens more than once, as stated above; the liquidity provider's balance should increase